scipy/stats/stats.py

```diff
@@ -2089,29 +2102,99 @@ def percentileofscore(a, score, kind='rank'):
     >>> stats.percentileofscore([1, 2, 3, 3, 4], 3, kind='mean')
     60.0
 
+    Multidimensional score arrays are supported:
+
+    >>> stats.percentileofscore([1, 2, 3, 3, 4], [2, 3])
+    array([40., 70.])
+
+    The inputs can be infinite:
+
+    >>> stats.percentileofscore([-np.inf, 0, 1, np.inf], [1, 100])
+    array([75., 75.])
+
+    If `a` is empty, or containing only nan's (which are omitted),
+    the resulting percentiles are all nan:
+
+    >>> stats.percentileofscore([], 1)
+    nan
+    >>> stats.percentileofscore([np.nan, 1, 2], [1, 2], nan_policy='omit')
+    array([ 50., 100.])
+    >>> stats.percentileofscore([np.nan, np.nan], [1, 2], nan_policy='omit')
+    array([nan, nan])
+
+    The 'omit' policy does not make sense if there are nan's among the scores,
+    and so an error in that case.
     """
-    if np.isnan(score):
-        return np.nan
+
     a = np.asarray(a)
     n = len(a)
+    score = np.asarray(score)
+
+    # Nan treatment
+    cna, npa = _contains_nan(a, nan_policy, summary_check=False)
+    cns, nps = _contains_nan(score, nan_policy, summary_check=False)
+
+    if (cna or cns) and nan_policy == 'raise':
+        raise ValueError("The input contains nan values")
+
+    elif cns and nan_policy == "omit":
+        raise ValueError("The input scores contains nan values,"
+                         " which is incompatible with the 'omit' policy.")
+
+    # If a score is nan, then the output should be nan
+    if cns and nan_policy == "propagate":
+        score = ma.masked_where(np.isnan(score), score)
+
+    # Don't count nans
+    if cna and nan_policy == "omit":
+        a = ma.masked_where(np.isnan(a), a)
+        n = a.count()
+
+    # With `propagate`, an `a` with nans should always yield nan
+    if cna and nan_policy == "propagate":
+        n = 0
+
+    # Cannot compare to empty list ==> nan
     if n == 0:
-        return 100.0
-
-    if kind == 'rank':
-        left = np.count_nonzero(a < score)
-        right = np.count_nonzero(a <= score)
-        pct = (right + left + (1 if right > left else 0)) * 50.0/n
```

Comment:
Just to check my understanding of the changes:

- `return` is delayed until the end so NaNs can be added back for implementing `nan_policy='propagate'`
- `np.count_nonzero` is applied along `axis=-1`
- the logic in this line `pct = (right + left + (1 if right > left else 0)) * 50.0/n` is vectorized

The rest - which makes it _appear_ to be a rewrite - is refactoring?

Reply:
Yes, correct on all counts.

Reply:
Well, I guess two points was left out:

- `nan_policy` has been implemented.
- More tests

These two points required by far the most work

Reply:
I just meant about this part of the code. I asked because the diff makes it look like it was rewritten, but really there are only these changes and the rest is refactoring. I agree that the refactoring makes it look cleaner, so OK to keep it here (but in general I'd recommend refactoring sparingly to speed review).

Reply:
Ok, thanks for the pointers 👍